File: src/data_structure/linked_list/insert_double.java
/*
  Insert Node at the end of a linked list 
  head pointer input could be NULL as well for empty list
  Node is defined as 
  class Node {
     int data;
     Node next;
     Node prev;
  }
*/

Node SortedInsert(Node head, int data) {
    Node node = new Node();
    node.data = data;

    if (head == null) {
        node.next = null;
        node.prev = null;
        return node;
    } else if (head.data > data) {
        node.next = head;
        node.prev = null;
        head.prev = node;
        return node;
    } else if (head.next == null) {
        head.next = node;
        node.prev = head;
        node.next = null;
        return head;
    }
    Node pre = head;
    Node cur = head.next;    
    while (cur != null && cur.data < data) {
        pre = cur;
        cur = cur.next;
    }
    pre.next = node;
    node.prev = pre;
    node.next = cur;
    if (cur != null) {
        cur.prev = node;
    }
    return head;
}

File: src/data_structure/linked_list/insert_double.cpp
/*
    Insert Node in a doubly sorted linked list 
    After each insertion, the list should be sorted
   Node is defined as
   struct Node
   {
     int data;
     Node *next;
     Node *prev
   }
*/

Node* SortedInsert(Node *head, int data)
{
    Node *new_node = (struct Node *) malloc(sizeof(struct Node));
    new_node->data = data;
    if (head == NULL) {
        new_node->next = NULL;
        new_node->prev = NULL;
        return new_node;
    } else if (data < head->data) {
        new_node->next = head;
        new_node->prev = NULL;
        head->prev = new_node;
        return new_node;
    } else if (head->next == NULL) {
        head->next = new_node;
        new_node->next = NULL;
        new_node->prev = head;
        return head;
    }
    Node *pre = head, *cur = head->next;
    while (cur != NULL) {
        if (cur->data > data) {
            pre->next = new_node;
            cur->prev = new_node;
            new_node->next = cur;
            new_node->prev = pre;
            return head;
        }
        pre = cur;
        cur = cur->next;
    }
    pre->next = new_node;
    new_node->next = NULL;
    new_node->prev = pre;
    return head;
}

